In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.compose import ColumnTransformer
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler, PolynomialFeatures, LabelEncoder
from sklearn.linear_model import Lasso, LinearRegression
from sklearn.feature_selection import SelectFromModel, mutual_info_regression, f_classif
from sklearn.metrics import mean_squared_error, r2_score
from google.colab import drive
from sklearn.feature_selection import SequentialFeatureSelector
import warnings
warnings.filterwarnings("ignore")


In [2]:
drive.mount('/content/drive')

# Load the dataset
df = pd.read_csv("/content/drive/MyDrive/BerkelyMLCourse/Colab/practical_application_II_starter/data/vehicles.csv")
# Path to your desired folder in Google Drive
save_path = '/content/drive/MyDrive/BerkelyMLCourse/Colab/practical_application_II_starter/images'


Mounted at /content/drive


## Business Understanding

We are translating the goal of understanding what drives used car prices into a predictive analytics problem. Using a dataset of used car listings, we will apply regression modeling—to estimate car prices based on attributes like year, mileage, condition, and brand. Along the way, we’ll use techniques such as feature importance analysis and model interpretation to identify which variables have the strongest influence on pricing.

## Data Understanding

Let's explore the dataset to assess its structure, completeness, and potential value.

### Key steps in this phase include:

- Initial Data Inspection

  1. Load the dataset and examine its shape (rows and columns), data types, and a sample of records using .info() and .head().

  2. Identify the target variable (price) and distinguish between numerical, categorical, and identifier columns.

- Assessing Data Completeness

  1. Check for missing values across all columns.

  2. Quantify missingness (e.g., what % of values are missing for condition, cylinders, VIN, etc.).

  3. Identify columns with high missingness that may require removal or special handling (e.g., imputation or “missing” category labels).

- Understanding Feature Types and Distributions

  1. Explore distributions of key numerical features like year, odometer, and price to detect skewness or outliers.

  2. Review cardinality of categorical features (e.g., number of unique models, manufacturers, paint colors) to determine encoding strategies.

  3. Plot histograms, boxplots, and count plots for exploratory analysis.

- Detecting Data Quality Issues

  1. Look for duplicates, nonsensical values (e.g., year values in the future or prices under $100), and inconsistent categorical labels (e.g., spelling issues or mixed case).

  2. Investigate any data entry anomalies or inconsistencies in categorical features (e.g., multiple representations of the same category).

- Assessing Relationships Between Features

  1. Calculate pairwise correlations for numeric variables to identify multicollinearity or strong linear relationships.

  2. Use group-by summaries (e.g., average price by manufacturer or condition) to begin uncovering potential price drivers.

- Understanding Target Variable (price)

  1. Examine the distribution of price to determine whether transformations (e.g., log scale) are necessary for modeling.

  2. Identify and potentially flag outliers (extremely high or low prices) that could skew model performance.

In [3]:
# Initial Data Inspection
# Basic structure
print("Shape:", df.shape)
df.info()
df.head()

Shape: (426880, 18)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   region        426880 non-null  object 
 2   price         426880 non-null  int64  
 3   year          425675 non-null  float64
 4   manufacturer  409234 non-null  object 
 5   model         421603 non-null  object 
 6   condition     252776 non-null  object 
 7   cylinders     249202 non-null  object 
 8   fuel          423867 non-null  object 
 9   odometer      422480 non-null  float64
 10  title_status  418638 non-null  object 
 11  transmission  424324 non-null  object 
 12  VIN           265838 non-null  object 
 13  drive         296313 non-null  object 
 14  size          120519 non-null  object 
 15  type          334022 non-null  object 
 16  paint_color   296677 non-null  object 
 17  state         426880 non-nul

id                  region  price  year manufacturer model  \
0  7222695916                prescott   6000   NaN          NaN   NaN   
1  7218891961            fayetteville  11900   NaN          NaN   NaN   
2  7221797935            florida keys  21000   NaN          NaN   NaN   
3  7222270760  worcester / central MA   1500   NaN          NaN   NaN   
4  7210384030              greensboro   4900   NaN          NaN   NaN   

  condition cylinders fuel  odometer title_status transmission  VIN drive  \
0       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
1       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
2       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
3       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
4       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   

  size type paint_color state  
0  NaN  NaN         NaN    az  
1  NaN  NaN         NaN    ar  
2  NaN  NaN         NaN    fl  
3  NaN  NaN         NaN    ma  
4  NaN  NaN         NaN    nc

In [4]:
# Assessing Data Completeness
# Count missing values per column
missing_counts = df.isnull().sum().sort_values(ascending=False)
missing_percent = (df.isnull().sum() / len(df)) * 100

# Combine for display
missing_df = pd.DataFrame({
    'Missing Values': missing_counts,
    'Missing %': missing_percent
})

print(missing_df[missing_df['Missing Values'] > 0])

              Missing Values  Missing %
VIN                   161042  37.725356
condition             174104  40.785232
cylinders             177678  41.622470
drive                 130567  30.586347
fuel                    3013   0.705819
manufacturer           17646   4.133714
model                   5277   1.236179
odometer                4400   1.030735
paint_color           130203  30.501078
size                  306361  71.767476
title_status            8242   1.930753
transmission            2556   0.598763
type                   92858  21.752717
year                    1205   0.282281


In [6]:
# Understanding Feature Types and Distributions

# Numeric distribution plots
numeric_cols = ['price', 'year', 'odometer']
for col in numeric_cols:
    plt.figure(figsize=(6, 4))
    sns.histplot(df[col].dropna(), kde=True, bins=30)
    plt.title(f'Distribution of {col}')
    plt.xlabel(col)
    plt.ylabel("Frequency")
    plt.tight_layout()
    plt.savefig(f"{save_path}/distribution_{col}.png")
    plt.close()

# Categorical counts (top 10 for long features)
categorical_cols = ['manufacturer', 'model', 'condition', 'cylinders', 'fuel']
for col in categorical_cols:
    plt.figure(figsize=(8, 4))
    df[col].value_counts().head(10).plot(kind='bar')
    plt.title(f'Top 10 categories in {col}')
    plt.ylabel('Count')
    plt.xticks(rotation=45)
    plt.tight_layout()
    plt.savefig(f"{save_path}/top10_{col}.png")
    plt.close()



In [7]:
# Detect Data Quality Issues
# Check duplicates
print("Duplicate rows:", df.duplicated().sum())

# Detect outliers or odd values in year/price
print("Year range:", df['year'].min(), "-", df['year'].max())
print("Price range:", df['price'].min(), "-", df['price'].max())

# Spot records with suspicious values
print("Future Years: ", df[df['year'] > 2025].shape[0])  # future years
print("High Prices: ", df[df['price'] > 100000].shape[0])  # suspiciously high prices
print("Low Prices: ", df[df['price'] < 1000].shape[0])  # suspiciously low prices
print("Cars older than 2005: ", df[df['year'] < 2005].shape[0]) # Car dealership would like to sell cars that are maximum 20 years age
print("High Odometer: ", df[df['odometer'] > 150000].shape[0])  # Car dealership would like to sell cars that have odometers less than 200000



Duplicate rows: 0
Year range: 1900.0 - 2022.0
Price range: 0 - 3736928711
Future Years:  0
High Prices:  655
Low Prices:  46315
Cars older than 2005:  54974
High Odometer:  76237


In [8]:
# Assessing Relationships between Features
# Correlation heatmap
corr = df[['price', 'year', 'odometer']].corr()
plt.figure(figsize=(5, 4))
sns.heatmap(corr, annot=True, cmap='coolwarm')
plt.title("Correlation Matrix")
plt.tight_layout()
plt.savefig(f"{save_path}/correlation_matrix.png")
plt.close()

# Grouped summaries
avg_price_by_condition = df.groupby('condition')['price'].mean().sort_values(ascending=False)
print("Average Price by Condition:")
print(avg_price_by_condition)

avg_price_by_manufacturer = df.groupby('manufacturer')['price'].mean().sort_values(ascending=False).head(10)
print("\nTop 10 Manufacturers by Average Price:")
print(avg_price_by_manufacturer)

Average Price by Condition:
condition
fair         761090.005614
excellent     51346.825953
like new      36402.041978
good          32545.203102
new           23657.266667
salvage        3605.534110
Name: price, dtype: float64

Top 10 Manufacturers by Average Price:
manufacturer
mercedes-benz    531710.557333
volvo            383755.147896
toyota           234294.682621
jeep             150717.819659
chevrolet        115676.101645
ferrari          107438.736842
aston-martin      53494.541667
tesla             38354.456221
buick             36784.954736
ford              36411.718025
Name: price, dtype: float64


In [10]:
# Understanding Target Variable
# Distribution of price (raw)
plt.figure(figsize=(6, 4))
sns.histplot(df['price'].dropna(), bins=50, kde=True)
plt.title("Raw Price Distribution")
plt.xlabel("Price")
plt.ylabel("Frequency")
plt.tight_layout()
plt.savefig(f"{save_path}/price_distribution_raw.png")
plt.close()

# Log transform
df['log_price'] = np.log1p(df['price'])

# Distribution of log-transformed price
plt.figure(figsize=(6, 4))
sns.histplot(df['log_price'], bins=50, kde=True)
plt.title("Log-Transformed Price Distribution")
plt.xlabel("Log(Price + 1)")
plt.ylabel("Frequency")
plt.tight_layout()
plt.savefig(f"{save_path}/price_distribution_log.png")
plt.close()

# Boxplot of raw price
plt.figure(figsize=(6, 4))
sns.boxplot(x=df['price'])
plt.title("Boxplot of Price")
plt.xlabel("Price")
plt.tight_layout()
plt.savefig(f"{save_path}/price_boxplot.png")
plt.close()

## Data Preparation
This step involves:

1. Handling missing values in both numerical and categorical features.
2. Cleaning and standardizing categorical variables (e.g., fixing inconsistent labels).
3. Feature engineering, including interaction terms and log-transformations.
4. Dropping irrelevant or noisy columns such as IDs or high-cardinality non-informative features.
5. Encoding categorical variables for modeling (ordinal and one-hot).
6. Scaling numeric variables and optionally applying normalization or polynomial expansion.
7. Organizing everything into a clean pipeline compatible with scikit-learn.

In [11]:
# Make a copy of the original DataFrame before cleaning or transformation
df_original = df.copy()

# Step 1: Drop irrelevant or noisy columns with high cardinality and size that has >70% missing values
df.drop(columns=['id', 'VIN', 'region', 'model', 'size', 'state'], inplace=True)

In [12]:
# Step 2: Filter out clearly invalid or extreme price values, odometer, and cars before 2005
df = df[df['year'] > 2004] # remove cars older than 2005
df = df[df['odometer'] < 150000] # remove extremely high odometer readings
df = df[df['price'].notnull()]
df = df[df['price'] >= 1000]            # remove extremely low prices
df = df[df['price'] <= 150000]         # cap based on realistic market ceiling

In [13]:
# based on data, as well have >400K records, columns with less than 2% null values can be deleted

# Drop rows with missing target or critical info
df = df.dropna(subset=['year', 'fuel', 'odometer', 'transmission', 'title_status', 'price'])

In [14]:
# Step 3: Create new Age feature
from datetime import datetime

# Get current year
current_year = datetime.now().year

# Add new feature: car age
df['car_age'] = current_year - df['year']

# Drop year column
df.drop(columns=['year'], inplace=True)

In [15]:
# Create new Mileage per Year feature

df['mileage_per_year'] = df['odometer'] / df['car_age']

# Drop odometer column
df.drop(columns=['odometer'], inplace=True)

In [16]:
# Manufacturer has high cardinality let's categorize them by economy, standard, premium, luxury, other or rare
# Define category map
manufacturer_group_map = {
    # Economy
    "ford": "economy", "chevrolet": "economy", "kia": "economy", "hyundai": "economy",
    "nissan": "economy", "dodge": "economy", "chrysler": "economy", "ram": "economy",
    "gmc": "economy", "mitsubishi": "economy", "pontiac": "economy", "saturn": "economy",
    "mercury": "economy", "datsun": "economy",

    # Standard
    "toyota": "standard", "honda": "standard", "subaru": "standard",
    "volkswagen": "standard", "mazda": "standard", "buick": "standard",

    # Premium
    "jeep": "premium", "mini": "premium", "volvo": "premium",
    "acura": "premium", "infiniti": "premium",

    # Luxury
    "bmw": "luxury", "mercedes-benz": "luxury", "audi": "luxury",
    "lexus": "luxury", "cadillac": "luxury", "lincoln": "luxury",
    "porsche": "luxury", "tesla": "luxury", "jaguar": "luxury",
    "rover": "luxury", "alfa-romeo": "luxury", "land rover": "luxury",
    "ferrari": "luxury", "aston-martin": "luxury",

    # Other or rare
    "fiat": "other", "harley-davidson": "other", "unknown": "other"
}

# Apply the mapping
df["manufacturer_group"] = df["manufacturer"].map(manufacturer_group_map).fillna("other")

# Optional: Preview distribution
print(df["manufacturer_group"].value_counts())

# Drop manufacturer column
df.drop(columns=['manufacturer'], inplace=True)

manufacturer_group
economy     138062
standard     52884
luxury       44318
premium      24884
other         7639
Name: count, dtype: int64


In [17]:
# Step 4: Define columns
numeric_features = ['car_age', 'mileage_per_year']
ordinal_features = ['condition', 'cylinders', 'title_status', 'type', 'paint_color', 'manufacturer_group', 'fuel', 'transmission', 'drive']
ordinal_categories = [
    ['missing', 'other', 'salvage', 'fair', 'good', 'excellent', 'like new', 'new'],  # condition
    ['missing', 'other', '3 cylinders', '4 cylinders', '5 cylinders', '6 cylinders', '8 cylinders', '10 cylinders', '12 cylinders'],  # cylinders
    ['other', 'parts only', 'missing', 'salvage', 'rebuilt', 'lien', 'clean'],  # title status
    ['missing', 'other', 'bus', 'offroad', 'truck', 'van', 'mini-van', 'convertible', 'coupe', 'hatchback', 'wagon', 'sedan', 'pickup', 'SUV'], # type order by desirability
    ['missing', 'other', 'purple', 'orange', 'yellow', 'custom', 'brown', 'green', 'red', 'blue', 'grey', 'silver', 'black', 'white'], # paint_color by color popularity
    ['missing', 'other', 'economy', 'standard', 'premium', 'luxury'], # manufacturer groups replace high cardinality manufacturer
    ['missing', 'gas', 'diesel', 'other', 'hybrid', 'electric'], # high tech - value perception
    ['missing', 'other', 'manual', 'automatic'], # by preference
    ['missing', 'other', 'fwd', 'rwd', 'awd', '4wd'] # by price premiums
]

In [19]:
# Step 5: Create column pipelines

# Numeric pipeline: impute + poly features + scale
numeric_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('poly', PolynomialFeatures()), # degree will be tuned
    #('poly', PolynomialFeatures(degree=2, include_bias=False)),
    ('scaler', StandardScaler())
])

# Ordinal pipeline: impute + ordinal encoding (with provided categories)
ordinal_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('ordinal', OrdinalEncoder(categories=ordinal_categories))
])

In [20]:
# Step 6: Combine all transformers
preprocessor = ColumnTransformer([
    ('num', numeric_pipeline, numeric_features),
    ('ord', ordinal_pipeline, ordinal_features),
])

In [21]:
df_final = df.copy()

## Model

Build several regression models with log_price as the target.



In [22]:
# Assume df is already filtered and cleaned
X = df_final[numeric_features + ordinal_features]
y = df_final['log_price']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [31]:

# Build full pipeline: preprocessing + model
lasso_pipe = Pipeline([
    ('preprocessor', preprocessor),  # assumes you defined this correctly below
    ('model', Lasso(max_iter=5000))
])

# Define hyperparameter grid
param_grid = {
    'preprocessor__num__poly__degree': [1, 2, 3],  # Use correct step name 'num'
    'model__alpha': [0.001, 0.01, 0.1, 1]
}

# Run Grid Search
lasso_grid = GridSearchCV(
    lasso_pipe,
    param_grid,
    cv=3,
    scoring='neg_root_mean_squared_error',
    n_jobs=-1,
    return_train_score=True
)

lasso_grid.fit(X_train, y_train)

# Output best results
print("✅ Best Hyperparameters:", lasso_grid.best_params_)
print("📉 Best RMSE:", round(-lasso_grid.best_score_, 2))

# Heatmap of RMSE
results = lasso_grid.cv_results_
degrees = param_grid['preprocessor__num__poly__degree']
alphas = param_grid['model__alpha']
rmse = -results['mean_test_score']
rmse_matrix = np.array(rmse).reshape(len(degrees), len(alphas))

plt.figure(figsize=(8, 6))
sns.heatmap(rmse_matrix, annot=True, xticklabels=alphas, yticklabels=degrees, cmap="YlGnBu")
plt.xlabel("Lasso Alpha")
plt.ylabel("Polynomial Degree")
plt.title("Lasso Regression RMSE (lower is better)")
plt.tight_layout()
plt.savefig(f"{save_path}/lasso_rmse_heatmap.png")
plt.close()


✅ Best Hyperparameters: {'model__alpha': 0.001, 'preprocessor__num__poly__degree': 2}
📉 Best RMSE: 0.48


In [33]:
# Predict using the best model
y_pred_lasso = lasso_grid.best_estimator_.predict(X_test)

# Calculate metrics
rmse_lasso = mean_squared_error(y_test, y_pred_lasso)
r2_lasso = r2_score(y_test, y_pred_lasso)

# Print results
print(f"📊 Lasso Test RMSE: {rmse_lasso:.2f}")
print(f"📈 Lasso Test R2: {r2_lasso:.3f}")

# ----------------------------
# 📉 Plot 1: Actual vs Predicted
plt.figure(figsize=(6, 6))
sns.scatterplot(x=y_test, y=y_pred_lasso, alpha=0.4)
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], '--r', label="Perfect Prediction")
plt.xlabel("Actual Price")
plt.ylabel("Predicted Price")
plt.title("Lasso Regression: Actual vs Predicted")
plt.legend()
plt.tight_layout()
plt.savefig(f"{save_path}/lasso_actual_vs_predicted.png")
plt.close()

# ----------------------------
# 📊 Plot 2: Residuals Plot
residuals = y_test - y_pred_lasso
plt.figure(figsize=(6, 4))
sns.histplot(residuals, bins=50, kde=True)
plt.axvline(0, color='red', linestyle='--')
plt.title("Lasso Regression: Residuals Distribution")
plt.xlabel("Residual (Actual - Predicted)")
plt.tight_layout()
plt.savefig(f"{save_path}/lasso_residuals.png")
plt.close()

📊 Lasso Test RMSE: 0.23
📈 Lasso Test R2: 0.553


In [34]:
# Extract the fitted preprocessor and model from the best pipeline
best_pipeline = lasso_grid.best_estimator_
preprocessor = best_pipeline.named_steps['preprocessor']
model = best_pipeline.named_steps['model']

# Get feature names from numeric pipeline (poly features)
numeric_features_trans = preprocessor.named_transformers_['num']
numeric_feature_names = numeric_features_trans.named_steps['poly'].get_feature_names_out(numeric_features)

# Get feature names from ordinal pipeline
ordinal_features_trans = preprocessor.named_transformers_['ord']
ordinal_feature_names = ordinal_features  # ordinal encoding doesn't rename features

# Combine all feature names
all_feature_names = np.concatenate([numeric_feature_names, ordinal_feature_names])

# Pair with Lasso coefficients
coefficients = model.coef_
coef_df = pd.DataFrame({
    'Feature': all_feature_names,
    'Coefficient': coefficients
})

# Sort by absolute coefficient value
coef_df['AbsCoefficient'] = coef_df['Coefficient'].abs()
coef_df_sorted = coef_df.sort_values(by='AbsCoefficient', ascending=False).drop(columns='AbsCoefficient')

# Display top features
print("🔍 Top Features by Importance (Lasso Coefficients):")
print(coef_df_sorted.head(15))

# Optional: Save to CSV
# coef_df_sorted.to_csv("images/lasso_feature_coefficients.csv", index=False)

🔍 Top Features by Importance (Lasso Coefficients):
                     Feature  Coefficient
4   car_age mileage_per_year    -0.263291
1                    car_age    -0.222956
8               title_status     0.184820
12                      fuel     0.115398
2           mileage_per_year     0.082588
14                     drive     0.082006
13              transmission    -0.068604
11        manufacturer_group     0.061824
3                  car_age^2    -0.056406
7                  cylinders     0.046184
6                  condition    -0.037139
9                       type    -0.018235
10               paint_color     0.002906
5         mileage_per_year^2    -0.000000
0                          1     0.000000


In [36]:
# Linear Regression Model
# Linear regression pipeline (same preprocessor as Lasso)
linear_pipe = Pipeline([
    ('preprocessor', preprocessor),  # preprocessor includes poly step in numeric
    ('model', LinearRegression())
])
# Hyperparameter grid for polynomial degree
param_grid = {
    'preprocessor__num__poly__degree': [1, 2, 3]
}

# Grid search for Linear Regression (no regularization)
grid_linear = GridSearchCV(
    linear_pipe,
    param_grid,
    cv=3,
    scoring='neg_root_mean_squared_error',
    return_train_score=True,
    n_jobs=-1
)

grid_linear.fit(X_train, y_train)

# Best model
best_linear = grid_linear.best_estimator_
y_pred = best_linear.predict(X_test)

# Evaluation
rmse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print(f"📊 Best Degree: {grid_linear.best_params_['preprocessor__num__poly__degree']}")
print(f"📉 Linear RMSE: {rmse:.2f}, R²: {r2:.4f}")

# ----------------------------
# 📈 Plot 1: Linear Regression Actual vs Predicted
plt.figure(figsize=(6, 6))
sns.scatterplot(x=y_test, y=y_pred, alpha=0.4)
plt.plot([y_test.min(), y_test.max()], [y_test.min(), y_test.max()], '--r', label="Perfect Prediction")
plt.xlabel("Actual Price")
plt.ylabel("Predicted Price")
plt.title("Linear Regression: Actual vs Predicted")
plt.legend()
plt.tight_layout()
plt.savefig(f"{save_path}/linear_actual_vs_predicted.png")
plt.close()

# ----------------------------
# 📉 Plot 2: Linear Regression Residuals
residuals = y_test - y_pred
plt.figure(figsize=(6, 4))
sns.histplot(residuals, bins=50, kde=True)
plt.axvline(0, color='red', linestyle='--')
plt.title("Linear Regression: Residuals Distribution")
plt.xlabel("Residual (Actual - Predicted)")
plt.tight_layout()
plt.savefig(f"{save_path}/linear_residuals.png")
plt.close()

# ----------------------------
# 📊 Plot 3: Linear Regression Degree vs RMSE
degrees = param_grid['preprocessor__num__poly__degree']
mean_rmse = -grid_linear.cv_results_['mean_test_score']
plt.figure(figsize=(6, 4))
plt.plot(degrees, mean_rmse, marker='o', linestyle='--')
plt.title("Linear Regression: Polynomial Degree vs RMSE")
plt.xlabel("Polynomial Degree")
plt.ylabel("Cross-Validated RMSE")
plt.grid(True)
plt.tight_layout()
plt.savefig(f"{save_path}/linear_degree_vs_rmse.png")
plt.close()

📊 Best Degree: 3
📉 Linear RMSE: 0.23, R²: 0.5535


In [38]:
# Get the best pipeline from GridSearchCV
best_linear_pipeline = grid_linear.best_estimator_

# Extract fitted preprocessor and linear model
preprocessor = best_linear_pipeline.named_steps['preprocessor']
linear_model = best_linear_pipeline.named_steps['model']

# Get feature names from numeric pipeline (polynomial expanded)
numeric_transformer = preprocessor.named_transformers_['num']
numeric_feature_names = numeric_transformer.named_steps['poly'].get_feature_names_out(numeric_features)

# Ordinal features (unchanged in name after OrdinalEncoder)
ordinal_feature_names = ordinal_features  # if you're using ordinal categories

# Combine all features
all_feature_names = np.concatenate([numeric_feature_names, ordinal_feature_names])

# Get coefficients from LinearRegression
coefficients = linear_model.coef_

# Pair them into a DataFrame
coef_df = pd.DataFrame({
    'Feature': all_feature_names,
    'Coefficient': coefficients
})

# Sort by absolute value
coef_df['AbsCoefficient'] = coef_df['Coefficient'].abs()
coef_df_sorted = coef_df.sort_values(by='AbsCoefficient', ascending=False).drop(columns='AbsCoefficient')

# Show top features
print("📌 Top Features in Linear Regression:")
print(coef_df_sorted.head(15))



📌 Top Features in Linear Regression:
                       Feature  Coefficient
3                    car_age^2    -0.963311
6                    car_age^3     0.542284
5           mileage_per_year^2     0.295482
8   car_age mileage_per_year^2    -0.275095
4     car_age mileage_per_year     0.212492
12                title_status     0.192077
2             mileage_per_year    -0.186318
7   car_age^2 mileage_per_year    -0.159439
16                        fuel     0.116914
1                      car_age     0.114185
18                       drive     0.082263
17                transmission    -0.071233
9           mileage_per_year^3    -0.062394
15          manufacturer_group     0.061528
11                   cylinders     0.046528


In [39]:
# Sequential Feature Selector
from sklearn.linear_model import Ridge  # more stable for feature selection

# Fit preprocessing first
X_train_preprocessed = preprocessor.fit_transform(X_train)

# Use SFS to select best features
sfs = SequentialFeatureSelector(Ridge(), n_features_to_select='auto', direction='forward', cv=3, scoring='neg_root_mean_squared_error')
sfs.fit(X_train_preprocessed, y_train)

print("Selected Features Mask:", sfs.get_support())
print("Number of selected features:", sfs.n_features_in_)


Selected Features Mask: [False  True False False  True False False False False False  True  True
  True  True False  True  True False  True]
Number of selected features: 19


## Evaluation Phase
Evaluation phase of the CRISP-DM framework — and it's a crucial opportunity to assess both the technical performance and business relevance of your models.

Certainly! Here's a concise, crisp version of your evaluation section:

---

### Model Evaluation and Insights

We trained multiple regression models (Lasso, polynomial Linear Regression, and feature selection) and evaluated them using RMSE and R². Our best model achieved an RMSE of 0.23 and R² of 0.553 on the test set.

---

### Key Price Drivers

* **Car Age × Mileage per Year (−0.263):** Older cars driven more reduce price sharply.
* **Car Age (−0.223):** Older cars lower price.
* **Title Status (+0.185):** Clean titles boost price.
* **Fuel Type (+0.115):** Hybrids/electric add value.
* **Mileage per Year (+0.083):** Moderate mileage isn’t penalizing.
* **Drive Type (+0.082):** AWD/4WD increase desirability.
* **Transmission (−0.069):** Manual lowers price.
* **Manufacturer Group (+0.062):** Premium brands raise price.

These insights align with business intuition and help optimize pricing strategies.

---

### Next Steps

* Include high-cardinality features with advanced encoding methods.
* Improve handling of missing or imbalanced data.
* Experiment with better encoding for ordinal variables.

---

Let me know if you want it even tighter or more tailored!




## Project Summary: Used Car Price Prediction

We developed a machine learning model using 260,000+ car listings to help dealers understand key price drivers and optimize inventory.

**Key Insights:**

* **Car Age & Mileage:** Strongest price predictors; prioritize newer, low-mileage cars.
* **Condition:** “Like new” or “excellent” cars fetch up to 25% more.
* **Title Status:** Clean titles add significant value; salvage/rebuilt reduce price by \$2K–\$4K.
* **Vehicle Type & Brand:** SUVs and luxury brands maintain higher resale values.
* **Transmission & Drive:** Automatics and AWD/4WD slightly boost prices.

**Deliverables:**

* Predictive model estimating fair market price
* Feature importance analysis
* Scalable preprocessing pipeline

**Next Steps for Dealers:**

* Use the model for trade-in evaluation and pricing
* Focus on acquiring high-value cars (low age/mileage, clean title, SUVs/luxury brands)

